x-pack/plugins/security_solution/server/endpoint/services/artifacts/manifest_manager/manifest_manager.ts

```diff
@@ -835,4 +855,231 @@ export class ManifestManager {
+    if (!this.cachedUnifiedManifestsSO.length) {
+      this.cachedUnifiedManifestsSO =
+        await this.getUnifiedManifestClient().getAllUnifiedManifests();
+    }
+    return this.cachedUnifiedManifestsSO;
+  }
+
+  protected clearCachedUnifiedManifestsSO(): void {
+    // Clear cached Unified Manifests after successful commit.
+    this.cachedUnifiedManifestsSO = [];
+  }
+
+  public transformUnifiedManifestSOtoLegacyManifestSO(
+    unifiedManifestsSo: InternalUnifiedManifestSchema[]
+  ): {
+    version: string;
+    attributes: {
+      artifacts: Array<
+        { artifactId: string; policyId: undefined } | { artifactId: string; policyId: string }
+      >;
+      semanticVersion: string;
+      schemaVersion: ManifestSchemaVersion;
+    };
+  } {
+    const globalUnifiedManifest = unifiedManifestsSo.find((a) => a.policyId === '.global');
+    return {
+      version: 'WzQ3NzAsMV0=', // version is hardcoded since it was used only to determine whether to create a new manifest or update an existing one
+      attributes: {
+        artifacts: [
+          ...(globalUnifiedManifest?.artifactIds.map((artifactId) => ({
+            artifactId,
+            policyId: undefined,
+          })) ?? []),
+          ...unifiedManifestsSo.reduce(
+            (acc: Array<{ artifactId: string; policyId: string }>, unifiedManifest) => {
+              if (unifiedManifest.policyId === '.global') {
+                return acc;
+              }
+              return [
+                ...acc,
+                ...unifiedManifest.artifactIds.map((artifactId) => ({
+                  policyId: unifiedManifest.policyId,
+                  artifactId,
+                })),
+              ];
+            },
+            []
+          ),
+        ],
+        semanticVersion: globalUnifiedManifest?.semanticVersion ?? '1.0.0',
+        schemaVersion: this.schemaVersion,
+      },
+    };
+  }
+
+  public transformLegacyManifestSOtoUnifiedManifestSO(
+    manifestSo: InternalManifestSchema,
+    unifiedManifestsSo: InternalUnifiedManifestSchema[]
+  ): Array<Omit<InternalUnifiedManifestUpdateSchema, 'id'> & { id?: string }> {
+    return manifestSo.artifacts.reduce(
+      (
+        acc: Array<InternalUnifiedManifestBaseSchema & { id?: string }>,
+        { artifactId, policyId = '.global' }
+      ) => {
+        const existingPolicy = acc.find((item) => item.policyId === policyId);
+        if (existingPolicy) {
+          existingPolicy.artifactIds.push(artifactId);
+        } else {
+          const existingUnifiedManifestSo = unifiedManifestsSo.find(
+            (item) => item.policyId === policyId
+          );
+
+          acc.push({
+            policyId,
+            artifactIds: [artifactId],
+            semanticVersion: existingUnifiedManifestSo?.semanticVersion ?? '1.0.0',
+            id: existingUnifiedManifestSo?.id,
+          });
+        }
+        return acc;
+      },
+      []
+    );
+  }
```

Comment:
are we good here with this quadratic time complexity?

update: measured it on my test env, took `10 ms`  this way, `3 ms` with optimization.
if you have time for it,  I'd suggest to improve it by building the output as an object where `policId`s are the keys, and afterwads converting the object to an array
```suggestion
    const manifestObject = manifestSo.artifacts.reduce(
      (
        acc: Record<string, InternalUnifiedManifestBaseSchema & { id?: string }>,
        { artifactId, policyId = '.global' }
      ) => {
        const existingPolicy = acc[policyId];
        if (existingPolicy) {
          existingPolicy.artifactIds.push(artifactId);
        } else {
          const existingUnifiedManifestSo = unifiedManifestsSo.find(
            (item) => item.policyId === policyId
          );

          acc[policyId] = {
            policyId,
            artifactIds: [artifactId],
            semanticVersion: existingUnifiedManifestSo?.semanticVersion ?? '1.0.0',
            id: existingUnifiedManifestSo?.id,
          };
        }
        return acc;
      },
      {}
    );
    return Object.values(manifestObject);
```

Reply:
Great input, thank you! 

95f28a9b1a2412e59839530b76946606dc87d0ff